controllers/machinedeletion_controller.go

```diff
@@ -63,19 +73,54 @@ func (r *MachineDeletionReconciler) Reconcile(ctx context.Context, req ctrl.Requ
 	if remediation = r.getRemediation(ctx, req); remediation == nil {
 		return ctrl.Result{}, nil
 	}
-	//not a machine based remediation
-	machineOwnerRef := getMachineOwnerRef(remediation)
-	if machineOwnerRef == nil {
-		return ctrl.Result{}, nil
+
+	var machine *unstructured.Unstructured
+	//Health check was done by NHC
+	if node, err := r.getNodeFromMdr(remediation); err == nil {
+		if machine, err = r.buildMachineFromNode(node); err != nil {
+			return ctrl.Result{}, err
+		}
+		if isMachineBelongToMasterNode(machine) {
+			return ctrl.Result{}, nil
+		}
+
+	} else { //Failed both in fetching the machine and in fetching the node
+		return ctrl.Result{}, err
 	}
-	//delete the machine
-	if err := r.deleteMachine(ctx, buildMachine(machineOwnerRef, remediation)); err != nil {
+
+	if err := r.deleteMachineOfNode(ctx, machine, remediation.Name); err != nil {
 		return ctrl.Result{}, err
 	}
 
 	return ctrl.Result{}, nil
 }
 
+func (r *MachineDeletionReconciler) deleteMachineOfNode(ctx context.Context, machine *unstructured.Unstructured, nodeName string) error {
+	//delete the machine
+	_ = r.Client.Delete(ctx, machine)
+
+	key := client.ObjectKey{
+		Name:      machine.GetName(),
+		Namespace: machine.GetNamespace(),
+	}
+
+	//verify machine is deleted
+	if err := r.Get(context.TODO(), key, machine); !errors.IsNotFound(err) {
+		return fmt.Errorf(failedToDeleteMachineError, nodeName)
+	}
+	return nil
+}
+
+func isMachineBelongToMasterNode(machine *unstructured.Unstructured) bool {
```

Comment:
the name of this function doesn't describe what it does.
it doesn't check if the argument is a master machine

Reply:
>it doesn't check if the argument is a master machine

That is true, but notice that this method suppose to check if the machine belongs to a master **node**.
The indication of master node is that it doesn't have a an OwnerRef of a MachineKind.

Reply:
> The indication of master node is that it doesn't have a an OwnerRef of a MachineKind.

there could be a machine without an ownerRef that is not a master.
The purpose of this check should not be if it's master or not. 
The check should be - will there by any controller that will re-create this machine after I delete it?

if we want to check for masters as well, that's fine, but should be done explicitly by checking the relevant label.